### PR TITLE
HRCPP-209 Added codec up to 2.4 without topology evaluation and events

### PR DESCRIPTION
**Pull request by rigazilla**
Added codec up to 2.4. Topology evaluation and event are still missing.

### CODE REVIEW

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity HotRod C++ Client :: Pull requests monitor - WIN_S2012 [Build 249](http://ci.infinispan.org/viewLog.html?buildId=33991&buildTypeId=HotRodCpp_PullRequestsMonitorWinS2012&guest=1) for 3c6cb97c55c29787b29ad4f286bb63f91c995ee6
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:00:02

Process exited with code 1

File: src/hotrod/impl/operations/PutIfAbsentOperation.cpp [hunk 1]
```diff
@@ -28,7 +28,7 @@ hrbytes PutIfAbsentOperation::executeOperation(Transport& transport)
     hrbytes previousValue;
     if (status == NO_ERROR_STATUS || status == NOT_PUT_REMOVED_REPLACED_STATUS) {
         previousValue =
-            AbstractKeyValueOperation<hrbytes>::returnPossiblePrevValue(transport);
+            AbstractKeyValueOperation<hrbytes>::returnPossiblePrevValue(transport,status);
```
Comment (tristantarrant) [inline src/hotrod/impl/operations/PutIfAbsentOperation.cpp]: formatting: add an extra space after comma

File: src/hotrod/impl/operations/PutOperation.cpp [hunk 2]
```diff
@@ -28,12 +28,12 @@ hrbytes PutOperation::executeOperation(Transport& transport) {
...
     TRACEBYTES("value = ", value);
     uint8_t status = sendPutOperation(transport, PUT_REQUEST, PUT_RESPONSE);
-    if (status != NO_ERROR_STATUS) {
+    if (!HotRodConstants::isSuccess(status)) {
         std::ostringstream message;
         message << "Unexpected response status: " << status;
         throw InvalidResponseException(message.str());
     }
-    return AbstractKeyValueOperation<hrbytes>::returnPossiblePrevValue(transport);
+    return AbstractKeyValueOperation<hrbytes>::returnPossiblePrevValue(transport,status);
```
Comment (tristantarrant) [inline src/hotrod/impl/operations/PutOperation.cpp]: same as above

File: src/hotrod/impl/operations/RemoveOperation.cpp [hunk 3]
```diff
@@ -23,9 +23,9 @@ hrbytes RemoveOperation::executeOperation(Transport& transport)
     TRACE("Execute Remove(flags=%u)", flags);
     TRACEBYTES("key = ", key);
     hrbytes result;
-    sendKeyOperation(key,
+    uint8_t status=sendKeyOperation(key,
         transport, REMOVE_REQUEST, REMOVE_RESPONSE);
-    return AbstractKeyOperation<hrbytes>::returnPossiblePrevValue(transport);
+    return AbstractKeyOperation<hrbytes>::returnPossiblePrevValue(transport,status);
```
Comment (tristantarrant) [inline src/hotrod/impl/operations/RemoveOperation.cpp]: again

File: src/hotrod/impl/operations/ReplaceIfUnmodifiedOperation.cpp [hunk 4]
```diff
@@ -29,8 +29,7 @@ VersionedOperationResponse ReplaceIfUnmodifiedOperation::executeOperation(
 
     //2) write message body
     transport.writeArray(key);
-    transport.writeVInt(lifespan);
-    transport.writeVInt(maxIdle);
+    codec.writeExpirationParams(transport,lifespan,maxIdle);
```
Comment (tristantarrant) [inline src/hotrod/impl/operations/ReplaceIfUnmodifiedOperation.cpp]: spacing

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity HotRod C++ Client :: Pull requests monitor - WIN_S2012 [Build 252](http://ci.infinispan.org/viewLog.html?buildId=34029&buildTypeId=HotRodCpp_PullRequestsMonitorWinS2012&guest=1) for 3c6cb97c55c29787b29ad4f286bb63f91c995ee6
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:00:02

Process exited with code 1

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity HotRod C++ Client :: Pull requests monitor - WIN_S2012 [Build 253](http://ci.infinispan.org/viewLog.html?buildId=34049&buildTypeId=HotRodCpp_PullRequestsMonitorWinS2012&guest=1) for 6d8627c39ba8cc74e1e940a69d9923adea4f9c9c
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:00:04

Process exited with code 1

**Comment by rigazilla** [comment]
Fixed formatting and added some missing code

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 322](http://ci.infinispan.org/viewLog.html?buildId=33992&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 6d8627c39ba8cc74e1e940a69d9923adea4f9c9c
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:07:59

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 323](http://ci.infinispan.org/viewLog.html?buildId=33993&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 6d8627c39ba8cc74e1e940a69d9923adea4f9c9c
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:08:01

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 327](http://ci.infinispan.org/viewLog.html?buildId=34033&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 6d8627c39ba8cc74e1e940a69d9923adea4f9c9c
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:09:16

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 328](http://ci.infinispan.org/viewLog.html?buildId=34034&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 6d8627c39ba8cc74e1e940a69d9923adea4f9c9c
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:10:47

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 329](http://ci.infinispan.org/viewLog.html?buildId=34050&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 6d8627c39ba8cc74e1e940a69d9923adea4f9c9c
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:08:50

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 330](http://ci.infinispan.org/viewLog.html?buildId=34051&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 6d8627c39ba8cc74e1e940a69d9923adea4f9c9c
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:08:01

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity HotRod C++ Client :: Pull requests monitor - WIN_S2012 [Build 254](http://ci.infinispan.org/viewLog.html?buildId=34150&buildTypeId=HotRodCpp_PullRequestsMonitorWinS2012&guest=1) for 9fefb60c16d3924ac91af9bcb0fdf712ccf1eae7
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:00:05

Process exited with code 1

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 331](http://ci.infinispan.org/viewLog.html?buildId=34151&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 9fefb60c16d3924ac91af9bcb0fdf712ccf1eae7
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:08:49

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 332](http://ci.infinispan.org/viewLog.html?buildId=34152&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 9fefb60c16d3924ac91af9bcb0fdf712ccf1eae7
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:08:05

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 333](http://ci.infinispan.org/viewLog.html?buildId=34153&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 9fefb60c16d3924ac91af9bcb0fdf712ccf1eae7
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:08:05

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 334](http://ci.infinispan.org/viewLog.html?buildId=34154&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 9fefb60c16d3924ac91af9bcb0fdf712ccf1eae7
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:08:06

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity HotRod C++ Client :: Pull requests monitor - WIN_S2012 [Build 260](http://ci.infinispan.org/viewLog.html?buildId=34270&buildTypeId=HotRodCpp_PullRequestsMonitorWinS2012&guest=1) for 9fefb60c16d3924ac91af9bcb0fdf712ccf1eae7
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:00:02

Process exited with code 1

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 344](http://ci.infinispan.org/viewLog.html?buildId=34271&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 9fefb60c16d3924ac91af9bcb0fdf712ccf1eae7
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:08:21

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 345](http://ci.infinispan.org/viewLog.html?buildId=34272&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 9fefb60c16d3924ac91af9bcb0fdf712ccf1eae7
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:08:51

**Comment by tristantarrant** [comment]
Pushed thanks

**Comment by infinispanrelease** [comment]
**FAILURE** - TeamCity HotRod C++ Client :: Pull requests monitor - WIN_S2012 [Build 264](http://ci.infinispan.org/viewLog.html?buildId=34568&buildTypeId=HotRodCpp_PullRequestsMonitorWinS2012&guest=1) for 9fefb60c16d3924ac91af9bcb0fdf712ccf1eae7
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:00:02

Process exited with code 1

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 351](http://ci.infinispan.org/viewLog.html?buildId=34570&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 9fefb60c16d3924ac91af9bcb0fdf712ccf1eae7
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:08:16

**Comment by infinispanrelease** [comment]
**SUCCESS** - TeamCity HotRod C++ Client :: Pull requests monitor - RHEL [Build 352](http://ci.infinispan.org/viewLog.html?buildId=34571&buildTypeId=HotRodCpp_PullRequestsMonitorRhel6&guest=1) for 9fefb60c16d3924ac91af9bcb0fdf712ccf1eae7
Tests: 0, 0 failed (0 new), 0 ignored. Build time: 00:08:09

**Pull request closed without merging**